File: two-pointers/two-sum-sorted/two-sum-sorted.java
// Two Pointers: Two Sum on a sorted array
//
// Not `public class`: a public class must live in a file named after it, and
// these files are kebab-case. The single-file launcher runs it either way.
import dsa.TestRunner;

class TwoSumSorted {

    static int[] twoSumSorted(int[] nums, int target) {
        int left = 0, right = nums.length - 1;
        while (left < right) {
            int sum = nums[left] + nums[right];
            if (sum == target) return new int[] { left, right };
            if (sum < target) left++;
            else right--;
        }
        return new int[] { -1, -1 };
    }

    public static void main(String[] args) {
        int[] nums = { 1, 2, 3, 4, 6 };
        int target = 6;

        int[] result = twoSumSorted(nums, target);
        System.out.println("Indices: " + result[0] + ", " + result[1]);

        // Every case in tests/two-sum-sorted.json, the same file the other languages read.
        // TestRunner and Json come from harness/java/out, which the run task puts on the classpath.
        TestRunner.runJsonCases("two-sum-sorted",
                in -> twoSumSorted(in.get(0).ints(), in.get(1).asInt()));

        TestRunner.runTests();
    }
}

File: two-pointers/two-sum-sorted/two-sum-sorted.cpp
// Two Pointers: Two Sum on a sorted array
#include <iostream>
#include <vector>

#include "test.hpp"  // harness/cpp is on the include path (see .vscode/tasks.json)

using namespace std;

vector<int> twoSumSorted(const vector<int>& nums, int target) {
    int left = 0, right = (int)nums.size() - 1;
    while (left < right) {
        int sum = nums[left] + nums[right];
        if (sum == target) return {left, right};
        if (sum < target) left++;
        else right--;
    }
    return {-1, -1};
}

int main() {
    vector<int> nums = {1, 2, 3, 4, 6};
    int target = 6;

    vector<int> result = twoSumSorted(nums, target);
    cout << "Indices: " << result[0] << ", " << result[1] << endl;

    // Every case in tests/two-sum-sorted.json, the same file the other languages read.
    dsa::runJsonCases("two-sum-sorted", [](const dsa::Json& in) {
        return twoSumSorted(in[0].ints(), (int)in[1].asInt());
    });

    return dsa::runTests();
}
